common/channel/src/libra_channel.rs

```diff
@@ -0,0 +1,151 @@
+use std::collections::{HashMap, VecDeque};
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::Waker;
+
+use futures::async_await::FusedStream;
+use futures::stream::Stream;
+use futures::task::Context;
+use futures::Poll;
+
+/// ValidatorMessage trait needs to be implemented by every message which gets pushed into the Libra
+/// channel so that we can ensure fairness among validators
+pub trait ValidatorMessage {
+    /// Extract the Validator from which this message arrived. This
+    /// will be used for ensuring fairness among validators.
+    fn get_validator(&self) -> &str;
+}
+
+/// Type is an enum which can be used as a configuration option for
+/// libra_channel
+pub enum Type {
+    LIFO,
+    FIFO,
+}
+
+struct PerValidatorQueue<T> {
+    queue_type: Type,
+    per_validator_queue: HashMap<String, VecDeque<T>>,
+    max_queue_size: usize,
+}
+
+impl<T: ValidatorMessage> PerValidatorQueue<T> {
+    fn new(queue_type: Type, max_queue_size: usize) -> Self {
```

Comment:
Could you please add a TODO for GC of old validator entries?

Reply:
I gave this some thought and I was not completely convinced that a GC mechanism is needed. The event processor should consume the queues for validators and if a validator dies, we will just have a couple of empty queues in a HashMap, which I think is fine. 

Maybe GC makes sense when the validators are highly dynamic and new validators keep coming and old validators keep dying.